### PR TITLE
Move classes from changedetection to common modules

### PR DESCRIPTION
**Pull request by ghale**
This is in preparation for using these classes in some of our maven libraries that currently duplicate this code.

Notes:
- `ClasspathResourceFilters` has been removed and the logic in `DefaultRuntimeClasspathNormalization` has been greatly simplified.
- `CachingResourceHasherTest` was testing functionality for both `CachingResourceHasher` and `DefaultResourceSnapshotterCacheService`.  Since only `CachingResourceHasher` was moved to normalization-java, that test was split into two tests, one for each class.

### CODE REVIEW

File: subprojects/core/src/main/java/org/gradle/normalization/internal/DefaultRuntimeClasspathNormalization.java [hunk 1]
```diff
@@ -110,14 +74,49 @@ public void ignoreManifest() {
...
+                                                .map(initializer)
+                                                .orElse(emptyValue);
+        }
+
+        T evaluate() {
+            if (value == null) {
+                value = valueSupplier.get();
+            }
+            return value;
+        }
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/normalization/internal/DefaultRuntimeClasspathNormalization.java:109]: Let's use `public` for methods even in `private` inner classes. That's what we do elsewhere, too. The rationale is that if you change the classes visibility to, say, `public`, these methods should also be visible with it.

File: subprojects/normalization-java/src/main/java/org/gradle/internal/fingerprint/classpath/impl/ClasspathFingerprintingStrategy.java [hunk 2]
```diff
@@ -181,7 +181,11 @@ private HashCode fingerprintTreeFile(RegularFileSnapshot fileSnapshot) {
             if (shouldBeIgnored) {
                 return null;
             }
-            return classpathResourceHasher.hash(fileSnapshot);
+            if (ZipHasher.isZipFile(fileSnapshot.getName())) {
+                return zipHasher.hash(fileSnapshot);
+            } else {
+                return classpathResourceHasher.hash(fileSnapshot);
+            }
```
Comment (lptr) [inline subprojects/normalization-java/src/main/java/org/gradle/internal/fingerprint/classpath/impl/ClasspathFingerprintingStrategy.java:188]: Shall we do this as a separate PR? The rest of the changes are quite unambiguous, but this one smells like it could introduce some breaking changes, and would be nice to keep track of separately.

File: subprojects/core/src/main/java/org/gradle/normalization/internal/DefaultRuntimeClasspathNormalization.java (hunk 1 above)
Comment (ghale) [reply subprojects/core/src/main/java/org/gradle/normalization/internal/DefaultRuntimeClasspathNormalization.java:109]: Fair enough - done.

File: subprojects/normalization-java/src/main/java/org/gradle/internal/fingerprint/classpath/impl/ClasspathFingerprintingStrategy.java (hunk 2 above)
Comment (ghale) [reply subprojects/normalization-java/src/main/java/org/gradle/internal/fingerprint/classpath/impl/ClasspathFingerprintingStrategy.java:188]: Good point.  I've moved these changes to https://github.com/gradle/gradle/pull/13352.
